torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,110 @@
+from collections import defaultdict
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+from torch.testing._internal.common_methods_invocations import (op_db, _DYNAMIC_DTYPES)
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           # all_types_and_half, # not exported by testing at the moment
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           # floating_types_and_half, # not exported by testing at the moment
+                           floating_types,
+                           integral_types,
+                           )
+
+DTYPE_DISPATCH = (
+    floating_types,
+    floating_and_complex_types,
+    integral_types,
+    all_types,
+    complex_types,
+    all_types_and_complex,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+if not TEST_CUDA:
+    warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA could be wrong")
+
+
+class OpInfoHelper:
+    def __init__(self, op_info):
+        self.op = op_info
+        self.sample_inputs_fn = op_info.sample_inputs_func
+
+    def get_supported_dtypes(self):
+        # Return a map of (device, supported_dtypes)
+        device_dtype = defaultdict(list)
+        for device in DEVICES:
+            for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
+                sample = self.sample_inputs_fn(self.op, device, dtype, False)[0]
+                try:
+                    self.op(sample.input, *sample.args, **sample.kwargs)
+                    device_dtype[device].append(dtype)
+                except RuntimeError as re:
+                    if "not implemented for" not in str(re):
+                        raise re
+                    pass
+
+        return device_dtype
+
+    def has_out_support(self, device, dtype):
+        try:
+            sample = self.sample_inputs_fn(None, device, dtype, False)[0]
+            op_out = partial(self.op, sample.input, *sample.args, **sample.kwargs)
+            op_out(out=torch.empty_like(sample.input))
```

Comment:
I would remove the out support detection from this first PR and let's just focus on dtypes

In addition to the above issues, I think this would need to run the operator and then produce out tensors that are "like" the output. Here the out tensor is "like" the input, which is true for unary elementwise operations but not most ops

Reply:
Forgot to do that :sweat_smile: . But I guess we should just skip adding `supports_out` for this version.